HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+    dvdnFunc_Sha_Adj : MargValueFunc2D
+        Marginal value function over normalized iliquid resources when the
+        agent is able to adjust his portfolio.
+    vFunc_Sha_Fxd : ValueFunc3D
+        Stage value function over normalized liquid resources, normalized
+        iliquid resources, and income contribution share when the agent is not
+        able to adjust his portfolio.
+    ShareFunc_Fxd : Interp3D
+        Income contribution share function over normalized liquid resources,
+        iliquid resources, and income contribution share when the agent is not
+        able to adjust his portfolio.
+        Should be an IdentityFunc.
+    dvdmFunc_Sha_Fxd : MargValueFunc3D
+        Marginal value function over normalized liquid resources when the agent
+        is not able to adjust his portfolio.
+    dvdnFunc_Sha_Fxd : MargValueFunc3D
+        Marginal value function over normalized iliquid resources when the
+        agent is not able to adjust his portfolio.
+    dvdsFunc_Sha_Fxd : Interp3D
+        Marginal value function over income contribution share when the agent
+        is not able to adjust his portfolio
+    """
+
+    distance_criteria = ["dvdmFunc_Sha_Adj", "dvdnFunc_Sha_Adj"]
+
+    def __init__(
+        self,
+        # Contribution stage, adjust
+        vFunc_Sha_Adj=None,
+        ShareFunc_Adj=None,
+        dvdmFunc_Sha_Adj=None,
+        dvdnFunc_Sha_Adj=None,
+        # Contribution stage, fixed
+        vFunc_Sha_Fxd=None,
+        ShareFunc_Fxd=None,
+        dvdmFunc_Sha_Fxd=None,
+        dvdnFunc_Sha_Fxd=None,
+        dvdsFunc_Sha_Fxd=None,
+    ):
+
+        # Contribution stage, adjust
+        if vFunc_Sha_Adj is None:
+            vFunc_Sha_Adj = NullFunc()
+        if ShareFunc_Adj is None:
+            ShareFunc_Adj = NullFunc()
+        if dvdmFunc_Sha_Adj is None:
+            dvdmFunc_Sha_Adj = NullFunc()
+        if dvdnFunc_Sha_Adj is None:
+            dvdnFunc_Sha_Adj = NullFunc()
+
+        # Contribution stage, fixed
+        if vFunc_Sha_Fxd is None:
+            vFunc_Sha_Fxd = NullFunc()
+        if ShareFunc_Fxd is None:
+            ShareFunc_Fxd = NullFunc()
+        if dvdmFunc_Sha_Fxd is None:
+            dvdmFunc_Sha_Fxd = NullFunc()
+        if dvdnFunc_Sha_Fxd is None:
+            dvdnFunc_Sha_Fxd = NullFunc()
+        if dvdsFunc_Sha_Fxd is None:
+            dvdsFunc_Sha_Fxd = NullFunc()
+
+        # Set attributes of self
+        self.vFunc_Sha_Adj = vFunc_Sha_Adj
+        self.ShareFunc_Adj = ShareFunc_Adj
+        self.dvdmFunc_Sha_Adj = dvdmFunc_Sha_Adj
+        self.dvdnFunc_Sha_Adj = dvdnFunc_Sha_Adj
+
+        self.vFunc_Sha_Fxd = vFunc_Sha_Fxd
+        self.ShareFunc_Fxd = ShareFunc_Fxd
+        self.dvdmFunc_Sha_Fxd = dvdmFunc_Sha_Fxd
+        self.dvdnFunc_Sha_Fxd = dvdnFunc_Sha_Fxd
+        self.dvdsFunc_Sha_Fxd = dvdsFunc_Sha_Fxd
+
+
+# Class for the consumption stage solution
+class RiskyContribCnsSolution(MetricObject):
```

Comment:
further in that logic... maybe the stage names should have longer, more legible names. But then appear less often, because the code is structured to use that information efficiently.

Reply:
I've deleted the redundant tags from class properties.

However, some degree of redundancy might help the reader. For instance, while solving the contribution share stage, I know that `v_func_next` will be of the `Cns` type, because the consumption stage comes next. However, somebody who is just figuring out what is going on in the model might thank calling the object `v_func_Cns_next` just so that he does not have to even think of it , or scroll up to the docstring.

Reply:
It's very hard to judge the aesthetics of these things since there is great variation in use of text editors, IDEs, documentation, etc.

I believe a more compact representation is standard across scientific libraries. If you examine the API documentation of many of the packages that we depend on, I believe you will come to see a convergence on a certain style.